isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java

```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
     List<RexNode> rexList =
-        project.getExpressions().stream()
+        project.getExpressions().stream().map(expr -> expr.accept(expressionRexConverter)).toList();
+
+    return relBuilder.push(child).project(rexList).build();
+  }
+
+  @Override
+  public RelNode visit(Join join) throws RuntimeException {
+    var left = join.getLeft().accept(this);
+    var right = join.getRight().accept(this);
+    var condition =
+        join.getCondition()
+            .map(c -> c.accept(expressionRexConverter))
+            .orElse(relBuilder.literal(true));
+    var joinType =
+        switch (join.getJoinType()) {
+          case INNER -> JoinRelType.INNER;
+          case LEFT -> JoinRelType.LEFT;
+          case RIGHT -> JoinRelType.RIGHT;
+          case OUTER -> JoinRelType.FULL;
+          case SEMI -> JoinRelType.SEMI;
+          case ANTI -> JoinRelType.ANTI;
+          case UNKNOWN -> throw new UnsupportedOperationException(
+              "Unknown join type is not supported");
+        };
+    return relBuilder.push(left).push(right).join(joinType, condition).build();
+  }
+
+  @Override
+  public RelNode visit(Aggregate aggregate) throws RuntimeException {
+    RelNode child = aggregate.getInput().accept(this);
+    var groupExprLists =
+        aggregate.getGroupings().stream()
             .map(
-                expr ->
-                    expr.accept(
-                        new ExpressionRexConverter(
-                            relOptCluster.getTypeFactory(), scalarFunctionConverter)))
+                gr ->
+                    gr.getExpressions().stream()
+                        .map(expr -> expr.accept(expressionRexConverter))
+                        .toList())
             .toList();
+    List<RexNode> groupExprs =
+        groupExprLists.stream().flatMap(Collection::stream).collect(Collectors.toList());
+    RelBuilder.GroupKey groupKey = relBuilder.groupKey(groupExprs, groupExprLists);
 
-    return relBuilder.push(child).project(rexList).build();
+    List<AggregateCall> aggregateCalls =
+        aggregate.getMeasures().stream().map(m -> fromMeasure(m)).toList();
+    return relBuilder.push(child).aggregate(groupKey, aggregateCalls).build();
+  }
+
+  private AggregateCall fromMeasure(Aggregate.Measure measure) {
+    var arguments =
+        measure.getFunction().arguments().stream()
+            .map(expr -> expr.accept(expressionRexConverter))
+            .toList();
+    var operator =
+        aggregateFunctionConverter.getSqlOperatorFromSubstraitFunc(
+            measure.getFunction().declaration().key(), measure.getFunction().outputType());
+    if (!operator.isPresent()) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Unable to find binding for call %s", measure.getFunction().declaration().name()));
+    }
+    List<Integer> argIndex = new ArrayList<>();
+    for (RexNode arg : arguments) {
+      if (!(arg instanceof RexInputRef)) {
```

Comment:
I think we need to fix this. In the case that there are compound expressions, I believe the rewrite to Calcite needs to be to insert a project with the calculation and refer to it from the aggregate function.

Reply:
Right, when Calcite converts from sql to RelNode, it would insert a project for compound expression.  The conversion of substrait -> calcite should process in a similar way, by adding a project.

However, for now, the substrait plan comes from Calcite RelNode, which has a project for compound expression.  In future, we would add substrait plan utility similar to Calcite's RelBuild/RexBuilder to generate arbitrary substrait plan/expr,  we are required to handle compound expression.  Without such substrait plan utility, it seems not straightforward to write a test with a substrait plan with compound expression in the agg call's argument.    

Shall we open a issue to track this gap and leave it as a TODO?

Reply:
That's fine but we should state in the exception that this just isn't implemented yet (as opposed to being impossible).

Reply:
Changed the message in the exception to "something is not implemented yet". 

Also, change the exception  from `UnsupportedOperationException`  to `IllegalArgumentException`, when the operator binding is not successful, or the operator is not the expected type.